✨ (apps:core) Implement soft-deletion

twidi opened the pull request:
We use `django-safedelete` to handle soft-deletion and
`django-postgres-extra` to manage unique indexes with a condition
(`WHERE deleted IS NULL`)

**Note to reviewers**: I know we don't have to test external libs. But we should at least test our "intentions", here the intentions are to avoid deleting content, and to ensure uniqueness. What to you think? Should I continue my work on these tests (not yet included in this PR)?

sampaccoud commented:
+1 for testing soft delete and uniqueness clauses even more so that you don't apply the same strategy on all models.

sampaccoud commented on marsha/core/tests/factories.py:
```diff
@@ -0,0 +1,93 @@
+    class Meta:  # noqa
+        abstract = True
+
+    deleted = False
```
Why do you need a factory for this if the default is None?

twidi replied:
I better enforce this. But I can remove this base factory if you want.

sampaccoud replied:
I think you should not enforce it in your tests because you might hide a bug (like a wrong default on deleted?). You should let it be the way it will be on a model when it is created in production...

IMO the good practice in factories is to either let the field have its default value or set a random value to prove that your tests work whatever the user sets... Faker provides realistic random values that a user might set for each type of field.

sampaccoud commented on marsha/core/tests/test_models.py:
```diff
@@ -0,0 +1,49 @@
+
+    def test_username_is_unique_even_if_deleted(self):
+        """Ensure that a username of a deleted user cannot be used by another."""
+        user = factories.UserFactory.create(username="foo")
```
why set username. It's the goal of the factory to do this. You can have a field in UserFactory:
```python
username = factory.Faker('username')
```

twidi replied:
It's for uniqueness, I force it to a specific value to use the same later to try to create another user with the same username.

twidi commented:
@sampaccoud, @jmaupetit: Updated with tests (and your typos corrected)

twidi commented:
Ok I can remove this, and no need for shell script as `django-admin test ...` exists

twidi commented:
@jmaupetit: removed arguments for `make test`, and explained how to use the django `test` command in the documentation

@sampaccoud: removed the base factory, using the default value of `deleted` (which was not `False`, by the way, but `None`)
